File: backend/src/main/java/com/ihecride/dao/EvaluationDAO.java
package com.ihecride.dao;

import com.ihecride.models.Evaluation;
import com.ihecride.utils.DBConnection;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class EvaluationDAO {

    public int insert(Evaluation e) {
        String sql = "INSERT INTO evaluations(trajet_id, evaluateur_id, evalue_id, note, commentaire) VALUES(?,?,?,?,?)";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            ps.setInt(1, e.getTrajetId());
            ps.setInt(2, e.getEvaluateurId());
            ps.setInt(3, e.getEvalueId());
            ps.setInt(4, e.getNote());
            ps.setString(5, e.getCommentaire());
            if (ps.executeUpdate() > 0) {
                try (ResultSet keys = ps.getGeneratedKeys()) {
                    if (keys.next()) return keys.getInt(1);
                }
            }
        } catch (SQLException ex) { ex.printStackTrace(); }
        return -1;
    }

    public List<Evaluation> findByEvalue(int evalueId) {
        List<Evaluation> list = new ArrayList<>();
        String sql = "SELECT e.*, CONCAT(u.prenom,' ',u.nom) AS evaluateur_nom FROM evaluations e JOIN users u ON e.evaluateur_id = u.id WHERE e.evalue_id = ? ORDER BY e.date DESC";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setInt(1, evalueId);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) list.add(map(rs));
            }
        } catch (SQLException e) { e.printStackTrace(); }
        return list;
    }

    public double moyenne(int userId) {
        String sql = "SELECT AVG(note) FROM evaluations WHERE evalue_id = ?";
        try (Connection c = DBConnection.getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setInt(1, userId);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) return rs.getDouble(1);
            }
        } catch (SQLException e) { e.printStackTrace(); }
        return 0.0;
    }

    private Evaluation map(ResultSet rs) throws SQLException {
        Evaluation e = new Evaluation();
        e.setId(rs.getInt("id"));
        e.setTrajetId(rs.getInt("trajet_id"));
        e.setEvaluateurId(rs.getInt("evaluateur_id"));
        e.setEvalueId(rs.getInt("evalue_id"));
        try { e.setEvaluateurNom(rs.getString("evaluateur_nom")); } catch (Exception ignored) {}
        e.setNote(rs.getInt("note"));
        e.setCommentaire(rs.getString("commentaire"));
        e.setDate(rs.getTimestamp("date"));
        return e;
    }
}

File: backend-php/dao/EvaluationDAO.php
<?php
/**
 * DAO : accès à la table evaluations.
 * Équivalent PHP de com.ihecride.dao.EvaluationDAO.
 */

require_once __DIR__ . '/../utils/DBConnection.php';
require_once __DIR__ . '/../models/Evaluation.php';

class EvaluationDAO {

    public function insert(Evaluation $e): int {
        $pdo = DBConnection::getConnection();
        try {
            $stmt = $pdo->prepare(
                "INSERT INTO evaluations(trajet_id, evaluateur_id, evalue_id, note, commentaire) VALUES(?,?,?,?,?)"
            );
            $ok = $stmt->execute([$e->trajetId, $e->evaluateurId, $e->evalueId, $e->note, $e->commentaire]);
            if ($ok && $stmt->rowCount() > 0) return (int) $pdo->lastInsertId();
        } catch (PDOException $ex) {
            error_log('EvaluationDAO::insert ' . $ex->getMessage());
        }
        return -1;
    }

    /** @return Evaluation[] */
    public function findByEvalue(int $evalueId): array {
        $sql = "SELECT e.*, CONCAT(u.prenom,' ',u.nom) AS evaluateur_nom
                FROM evaluations e JOIN users u ON e.evaluateur_id = u.id
                WHERE e.evalue_id = ? ORDER BY e.date DESC";
        $stmt = DBConnection::getConnection()->prepare($sql);
        $stmt->execute([$evalueId]);
        $list = [];
        foreach ($stmt->fetchAll() as $row) $list[] = $this->map($row);
        return $list;
    }

    public function moyenne(int $userId): float {
        $stmt = DBConnection::getConnection()->prepare(
            "SELECT AVG(note) AS avg FROM evaluations WHERE evalue_id = ?"
        );
        $stmt->execute([$userId]);
        $row = $stmt->fetch();
        return $row && $row['avg'] !== null ? (float) $row['avg'] : 0.0;
    }

    private function map(array $rs): Evaluation {
        $e = new Evaluation();
        $e->id             = isset($rs['id']) ? (int) $rs['id'] : null;
        $e->trajetId       = isset($rs['trajet_id']) ? (int) $rs['trajet_id'] : null;
        $e->evaluateurId   = isset($rs['evaluateur_id']) ? (int) $rs['evaluateur_id'] : null;
        $e->evalueId       = isset($rs['evalue_id']) ? (int) $rs['evalue_id'] : null;
        $e->evaluateurNom  = $rs['evaluateur_nom'] ?? null;
        $e->note           = isset($rs['note']) ? (int) $rs['note'] : null;
        $e->commentaire    = $rs['commentaire'] ?? null;
        $e->date           = $rs['date'] ?? null;
        return $e;
    }
}
